model/atmosphere/dycore/docs/_ext/icon4py_sphinx.py

```diff
@@ -0,0 +1,453 @@
+                                                    elif isinstance(current_node, ast.Name):
+                                                        call_chain.append(current_node.id)
+                                                        break
+                                                call_chain.reverse()
+                                                if call_chain[-1] == 'with_backend':
+                                                    # remove it from the call chain
+                                                    call_chain.pop()
+                                                if len(call_chain) == 1:
+                                                    # the method is imported directly, e.g.
+                                                    # from solve_nonhydro_program import stencil
+                                                    # self._stencil = stencil.with_backend(...)
+                                                    original_method_name = call_chain[0]
+                                                    if original_method_name in self.module.__dict__.keys() and not isinstance(getattr(self.module, original_method_name), types.ModuleType):
+                                                        # method_obj.definition_stage.definition.__module__
+                                                        method_obj = getattr(self.module, original_method_name)
+                                                        if type(method_obj).__module__.startswith('gt4py') and type(method_obj).__name__ == 'Program':
+                                                            # it's a decorated gt4py program
+                                                            module_full_name = method_obj.definition_stage.definition.__module__
+                                                            break
+                                                elif len(call_chain) >= 2:
+                                                    # the method is called from a module import, e.g.
+                                                    # import solve_nonhydro_program as nhsolve_prog
+                                                    # self._stencil = nhsolve_prog.stencil.with_backend(...)
+                                                    module_name = call_chain[0]
+                                                    original_method_name = call_chain[1]
+                                                    if module_name in self.module.__dict__.keys() and isinstance(getattr(self.module, module_name), types.ModuleType):
+                                                        module_obj = getattr(self.module, module_name)
+                                                        method_obj = getattr(module_obj, original_method_name)
+                                                        module_full_name = module_obj.__name__
+                                                        break
```

Comment:
Move all the statements related to the computation of `module_full_name` to a helper function.